### PR TITLE
Resolves #169 CredentialStore integration into subsystem

### LINKED ISSUE #169: CredentialStore integration into subsystem with example of how to obtain secret credential

### PR DESCRIPTION
**Pull request by pskopek**

### CODE REVIEW

File: src/main/java/org/wildfly/extension/elytron/CredentialStoreClient.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+import org.wildfly.security.password.interfaces.ClearPassword;
+
+import java.net.URISyntaxException;
+
+/**
+ * Client class used to talk to {@link CredentialStore} to obtain secret value (credential) from it.
+ *
+ * @author <a href="mailto:pskopek@redhat.com">Peter Skopek</a>
+ */
+public class CredentialStoreClient implements Destroyable {
```
Comment (darranl) [inline src/main/java/org/wildfly/extension/elytron/CredentialStoreClient.java]: In general almost nothing should be public within the subsystem, if there is something that needs to be used across the whole application server we may need to incorporate it in a suitable location within WildFly Core instead.

**Comment by darranl** [comment]
From the discussion we had during last years F2F meeting I think we were heading in the direction of using a complex attribute to reference the credential store - that way instead of a specially formatted string we have one attribute referencing the credential store and another to reference the alias within that credential store - that way capability and requirements based model time verification can at the very least verify that the named credential store does exist.

**Comment by pskopek** [comment]
Rebased and updated. It is ready to merge after https://github.com/wildfly/wildfly-core/pull/1816

**Comment by darranl** [comment]
Just chatted with Brian on this one, rather than trying to use the management schema as a common schema we think it will be safer if subsystems just duplicate the complex type definition - that would stop the build here failing.

**Comment by pskopek** [comment]
@darranl PR is fixed and I have duplicated complex type definition in this subsystem schema.

File: src/main/java/org/wildfly/extension/elytron/Capabilities.java [hunk 2]
```diff
@@ -57,6 +58,13 @@
     private static final String CAPABILITY_BASE = "org.wildfly.security.";
 
     static final String HTTP_AUTHENTICATION_FACTORY_CAPABILITY = CAPABILITY_BASE + "http-authentication-factory";
+    static final String CREDENTIAL_STORE_CLIENT_CAPABILITY = CAPABILITY_BASE + "credential-store-client";
+
+    static final RuntimeCapability<Void> CREDENTIAL_STORE_CLIENT_RUNTIME_CAPABILITY =  RuntimeCapability
+            .Builder.of(CREDENTIAL_STORE_CLIENT_CAPABILITY, true, CredentialStoreClient.class)
+            .build();
+
+    static final String HTTP_SERVER_AUTHENTICATION_CAPABILITY = CAPABILITY_BASE + "http-server-authentication";
```
Comment (darranl) [inline src/main/java/org/wildfly/extension/elytron/Capabilities.java]: I think this one is a rebase issue - we have now switched to using HTTP_AUTHENTICATION_FACTORY as the capability.

File: src/main/java/org/wildfly/extension/elytron/CredentialStoreAliasDefinition.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    static final StandardResourceDescriptionResolver RESOURCE_DESCRIPTION_RESOLVER = ElytronExtension.getResourceDescriptionResolver(ElytronDescriptionConstants.CREDENTIAL_STORE, ElytronDescriptionConstants.ALIAS);
+
+    static final SimpleAttributeDefinition SECRET_VALUE = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.SECRET_VALUE, ModelType.STRING)
+            .setStorageRuntime()
+            .setAllowNull(false)
+            .build();
+
+    private static final AttributeDefinition[] CONFIG_ATTRIBUTES = new AttributeDefinition[] {SECRET_VALUE, ENTRY_TYPE};
+    //private static final WriteAttributeHandler WRITE = new WriteAttributeHandler();
+
```
Comment (darranl) [inline src/main/java/org/wildfly/extension/elytron/CredentialStoreAliasDefinition.java]: As we need a clean up elsewhere probably remove this commented out line.

File: src/main/java/org/wildfly/extension/elytron/CredentialStoreAliasDefinition.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+                .setRemoveRestartLevel(OperationEntry.Flag.RESTART_NONE)
+                .setRuntime()
+        );
+    }
+
+    @Override
+    public void registerOperations(ManagementResourceRegistration resourceRegistration) {
+        super.registerOperations(resourceRegistration);
+        // We needed a custom add operation so we could specify the parameters.
+        resourceRegistration.registerOperationHandler(ADD_DEFINITION, ADD_HANDLER);
```
Comment (darranl) [inline src/main/java/org/wildfly/extension/elytron/CredentialStoreAliasDefinition.java]: As the ADD handler is static we should be able to add it to the paramers in the constructor.

File: src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+
+    private static class WriteAttributeHandler extends RestartParentWriteAttributeHandler {
+
+        WriteAttributeHandler() {
+            super(ElytronDescriptionConstants.CREDENTIAL_STORE, CONFIG_ATTRIBUTES);
+        }
+
+        @Override
+        protected ServiceName getParentServiceName(PathAddress arg0) {
+            return null;
```
Comment (darranl) [inline src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java]: Should we have a service name?

File: src/main/resources/schema/wildfly-elytron_1_0.xsd [hunk 6]
```diff
@@ -25,6 +25,8 @@
             attributeFormDefault="unqualified"
             version="1.0">
 
+    <xs:import namespace="urn:jboss:domain:5.0"/>
```
Comment (darranl) [inline src/main/resources/schema/wildfly-elytron_1_0.xsd]: This import can also be removed.

File: src/main/java/org/wildfly/extension/elytron/CredentialStoreAliasDefinition.java (hunk 4 above)
Comment (pskopek) [inline src/main/java/org/wildfly/extension/elytron/CredentialStoreAliasDefinition.java]: We cannot have operation definition in parameters. I am leaving this without change.

**Comment by pskopek** [comment]
Changes pushed.

**Pull request merged**